### PR TITLE
[haskell] support dynamic completion backend

### PR DESCRIPTION
**Pull request by d12frosted**

### CODE REVIEW

File: layers/+lang/haskell/funcs.el [hunk 1]
```diff
@@ -22,7 +22,59 @@
...
+    (`ghc-mod (spacemacs-haskell//setup-ghc-mod))))
+
+(defun spacemacs-haskell//setup-company ()
+  "Conditionally setup haskell backend."
+  (pcase haskell-completion-backend
+    (`ghci (spacemacs-haskell//setup-ghci-company))
+    (`intero (spacemacs-haskell//setup-intero-company))
+    (`dante (spacemacs-haskell//setup-dante-company))
+    (`ghc-mod (spacemacs-haskell//setup-ghc-mod-company))))
+
```
Comment (d12frosted) [inline layers/+lang/haskell/funcs.el]: This is not an empty line, it's `^L`.

File: layers/+lang/haskell/packages.el [hunk 2]
```diff
@@ -16,26 +16,19 @@
...
-     :requires company
-     :toggle (eq haskell-completion-backend 'intero))
+    (intero :requires company)
 
     ;; dante completion backend
     (dante
      :requires company
-     :toggle (and (version<= "25" emacs-version)
-                  (eq haskell-completion-backend 'dante)))
+     :toggle (version<= "25" emacs-version))
```
Comment (d12frosted) [inline layers/+lang/haskell/packages.el]: I haven't checked yet, but it looks rubbish that `dante` is available only on old Emacs.

File: layers/+lang/haskell/funcs.el (hunk 1 above)
Comment (sdwolfz) [reply layers/+lang/haskell/funcs.el]: :see_no_evil:

Comment (d12frosted) [reply layers/+lang/haskell/funcs.el]: 😸I was reading my PR and saw these double-empty lines. Was like - what the line?!

File: layers/+lang/haskell/packages.el (hunk 2 above)
Comment (d12frosted) [reply layers/+lang/haskell/packages.el]: I've checked it. And it is no longer the case. Removed all along.

**Comment by michaelpj** [comment]
It would be great to get this merged - it has a fairly trivial merge conflict.

I'm working on getting the lsp backend for Haskell working, and this refactoring is definitely desirable first.

**Comment by michaelpj** [comment]
I opened https://github.com/syl20bnr/spacemacs/pull/11887, which is based on this. I fixed the merge and removed the form-feed characters, but if this is merged then it should all work cleanly anyway.

**Comment by duianto** [comment]
The PR:
Haskell: add LSP support via HIE and lsp-haskell #11887
is now on the develop branch.

It seems to have been based on this PR.

**Comment by d12frosted** [comment]
Oh neat! Thank you both for doing this!

**Pull request closed without merging**